warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java

```diff
@@ -982,6 +982,161 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
         return assignments;
     }
 
+    /**
+     * <b>NOTE: Do NOT use this method with a rebuilding visitor. You cannot guarantee parentage will remain the same.</b>
+     * <p>
+     * Swaps the original node with the replacement, following rules for reference expressions, single terms and junctions
+     * <p>
+     * This method prevents reference expressions from having a single child.
+     * <p>
+     * This method also allows for merging similar junctions.
+     *
+     * @param original
+     *            the original node
+     * @param replacement
+     *            the replacement node
+     */
+    public static void replaceNodeSafely(JexlNode original, JexlNode replacement) {
+        JexlNode parent = original.jjtGetParent();
+
+        if (parent == null) {
+            return; // cannot do anything in this case
+        }
+
+        boolean multiValue = isJunction(replacement);
+        if (multiValue) {
+            replaceNodeWithMultiValue(parent, original, replacement);
+        } else {
+            replaceNodeWithSingleValue(parent, original, replacement);
+        }
+    }
+
+    /**
+     * This method accepts a multivalued replacement node (ASTAndNode or ASTOrNode) and attempts to merge it cleanly into the parent or grandparent node with a
+     * preference for a reference expression as the new parent.
+     *
+     * @param parent
+     *            the parent node
+     * @param original
+     *            the original node
+     * @param replacement
+     *            the replacement node
+     */
+    private static void replaceNodeWithMultiValue(JexlNode parent, JexlNode original, JexlNode replacement) {
+
+        // similar junctions, merge children
+        if (isEquivalentJunctions(parent, replacement)) {
+            mergeJunctions(parent, original, replacement);
+            return;
+        }
+
+        // if no 'legal' rewrite could occur with the parent, try the grandparent
+        JexlNode grandParent = parent.jjtGetParent();
+        if (grandParent != null) {
+            // reference expression, straight replacement
+            if (grandParent instanceof ASTReferenceExpression) {
+                JexlNodes.replaceChild(grandParent, parent, replacement);
+                return;
+            }
+
+            // similar junctions, merge children
+            if (isEquivalentJunctions(grandParent, replacement)) {
+                mergeJunctions(grandParent, parent, replacement);
+                return;
+            }
+        }
```

Comment:
does this need to consider the type of node of the parent since the method requires any JexlNode?

Reply:
No, it does not. The parent method can be used to replace reference expressions, and nodes, or nodes, or any leaf node. The check you see here simply ensure that any replacement produces a logically correct query tree, i.e., no adjacent junctions or reference expressions.